File: src/ch/deletescape/lawnchair/preferences/IconPackPreference.java
package ch.deletescape.lawnchair.preferences;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ch.deletescape.lawnchair.R;


public class IconPackPreference extends Preference {

    private final PackageManager pm;

    public IconPackPreference(Context context) {
        this(context, null);
    }

    public IconPackPreference(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public IconPackPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setLayoutResource(R.layout.preference_iconpack);
        pm = context.getPackageManager();
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        init();
        return super.onCreateView(parent);
    }

    private void init() {
        String currentPack = getPersistedString("");
        if (currentPack.isEmpty()) {
            setNone();
        } else {
            try {
                ApplicationInfo info = pm.getApplicationInfo(currentPack, 0);
                setIcon(info.loadIcon(pm));
                setSummary(info.loadLabel(pm));
            } catch (PackageManager.NameNotFoundException e) {
                setNone();
                persistString("");
            }
        }
    }

    private void setNone() {
        setIcon(getContext().getResources().getDrawable(R.mipmap.ic_launcher_home));
        setSummary("None");
    }

    @Override
    protected void onClick() {
        super.onClick();
        showDialog();
    }

    protected void showDialog() {
        final Map<String, IconPackInfo> packages = loadAvailableIconPacks();
        final IconAdapter adapter = new IconAdapter(getContext(), packages, getPersistedString(""));
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setAdapter(adapter, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int position) {
                String item = adapter.getItem(position);
                persistString(item);
                if (!item.isEmpty()) {
                    IconPackInfo packInfo = packages.get(item);
                    setIcon(packInfo.icon);
                    setSummary(packInfo.label);
                } else {
                    setNone();
                }
            }
        });
        builder.show();
    }

    private Map<String, IconPackInfo> loadAvailableIconPacks() {
        Map<String, IconPackInfo> iconPacks = new HashMap<>();
        List<ResolveInfo> list;
        list = pm.queryIntentActivities(new Intent("com.novalauncher.THEME"), 0);
        list.addAll(pm.queryIntentActivities(new Intent("org.adw.launcher.icons.ACTION_PICK_ICON"), 0));
        list.addAll(pm.queryIntentActivities(new Intent("com.dlto.atom.launcher.THEME"), 0));
        list.addAll(pm.queryIntentActivities(new Intent("android.intent.action.MAIN").addCategory("com.anddoes.launcher.THEME"), 0));
        for (ResolveInfo info : list) {
            iconPacks.put(info.activityInfo.packageName, new IconPackInfo(info, pm));
        }
        return iconPacks;
    }

    private static class IconPackInfo {
        String packageName;
        CharSequence label;
        Drawable icon;

        IconPackInfo(ResolveInfo r, PackageManager packageManager) {
            packageName = r.activityInfo.packageName;
            icon = r.loadIcon(packageManager);
            label = r.loadLabel(packageManager);
        }

        public IconPackInfo(String label, Drawable icon, String packageName) {
            this.label = label;
            this.icon = icon;
            this.packageName = packageName;
        }
    }

    private static class IconAdapter extends BaseAdapter {
        ArrayList<IconPackInfo> mSupportedPackages;
        LayoutInflater mLayoutInflater;
        String mCurrentIconPack;

        IconAdapter(Context context, Map<String, IconPackInfo> supportedPackages, String currentPack) {
            mLayoutInflater = LayoutInflater.from(context);
            mSupportedPackages = new ArrayList<>(supportedPackages.values());
            Collections.sort(mSupportedPackages, new Comparator<IconPackInfo>() {
                @Override
                public int compare(IconPackInfo lhs, IconPackInfo rhs) {
                    return lhs.label.toString().compareToIgnoreCase(rhs.label.toString());
                }
            });

            Resources res = context.getResources();
            String defaultLabel = "None";
            Drawable icon = res.getDrawable(R.mipmap.ic_launcher_home);
            mSupportedPackages.add(0, new IconPackInfo(defaultLabel, icon, ""));
            mCurrentIconPack = currentPack;
        }

        @Override
        public int getCount() {
            return mSupportedPackages.size();
        }

        @Override
        public String getItem(int position) {
            return mSupportedPackages.get(position).packageName;
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (convertView == null) {
                convertView = mLayoutInflater.inflate(R.layout.iconpack_dialog, parent);
            }
            IconPackInfo info = mSupportedPackages.get(position);
            TextView txtView = (TextView) convertView.findViewById(R.id.title);
            txtView.setText(info.label);
            ImageView imgView = (ImageView) convertView.findViewById(R.id.icon);
            imgView.setImageDrawable(info.icon);
            RadioButton radioButton = (RadioButton) convertView.findViewById(R.id.radio);
            radioButton.setChecked(info.packageName.equals(mCurrentIconPack));
            return convertView;
        }
    }

}

File: app/src/main/java/ch/deletescape/lawnchair/preferences/IconPackPreference.kt
package ch.deletescape.lawnchair.preferences

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.drawable.Drawable
import android.support.v7.preference.Preference
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import ch.deletescape.lawnchair.R
import ch.deletescape.lawnchair.Utilities
import java.util.*

class IconPackPreference @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : Preference(context, attrs, defStyleAttr) {

    private val pm: PackageManager

    init {
        layoutResource = R.layout.preference_iconpack
        pm = context.packageManager
    }

    override fun onAttached() {
        super.onAttached()
        init()
    }

    private fun init() {
        val currentPack = getPersistedString("")
        if (currentPack.isEmpty()) {
            setNone()
        } else {
            try {
                val info = pm.getApplicationInfo(currentPack, 0)
                icon = info.loadIcon(pm)
                summary = info.loadLabel(pm)
            } catch (e: PackageManager.NameNotFoundException) {
                setNone()
                persistString("")
            }

        }
    }

    private fun setNone() {
        icon = Utilities.getMyIcon(context)
        summary = "None"
    }

    override fun onClick() {
        super.onClick()

        // TODO: Add some 'Arr!' flavor to it
        if (Utilities.isBlacklistedAppInstalled(context)) {
            Toast.makeText(context, R.string.unauthorized_device, Toast.LENGTH_SHORT).show()
            return
        }

        showDialog()
    }

    private fun showDialog() {
        val packages = loadAvailableIconPacks()
        val adapter = IconAdapter(context, packages, getPersistedString(""))
        val builder = AlertDialog.Builder(context)
        builder.setAdapter(adapter) { _, position ->
            val item = adapter.getItem(position)
            persistString(item)
            if (!item.isEmpty()) {
                val packInfo = packages[item]
                icon = packInfo!!.icon
                summary = packInfo.label
            } else {
                setNone()
            }

            val alternativeIcons = Utilities.getAlternativeIconList(context);
            if (alternativeIcons.size > 0) {
                Utilities.showResetAlternativeIcons(context, alternativeIcons)
            }
        }
        builder.show()
    }

    private fun loadAvailableIconPacks(): Map<String, IconPackInfo> {
        val iconPacks = HashMap<String, IconPackInfo>()
        val list = pm.queryIntentActivities(Intent("com.novalauncher.THEME"), 0)
        list.addAll(pm.queryIntentActivities(Intent("org.adw.launcher.icons.ACTION_PICK_ICON"), 0))
        list.addAll(pm.queryIntentActivities(Intent("com.dlto.atom.launcher.THEME"), 0))
        list.addAll(pm.queryIntentActivities(Intent("android.intent.action.MAIN").addCategory("com.anddoes.launcher.THEME"), 0))
        for (info in list) {
            iconPacks.put(info.activityInfo.packageName, IconPackInfo(info, pm))
        }
        return iconPacks
    }

    private class IconPackInfo {
        internal var packageName: String
        internal var label: CharSequence
        internal var icon: Drawable

        internal constructor(r: ResolveInfo, packageManager: PackageManager) {
            packageName = r.activityInfo.packageName
            icon = r.loadIcon(packageManager)
            label = r.loadLabel(packageManager)
        }

        constructor(label: String, icon: Drawable, packageName: String) {
            this.label = label
            this.icon = icon
            this.packageName = packageName
        }
    }

    private class IconAdapter internal constructor(context: Context, supportedPackages: Map<String, IconPackInfo>, internal var mCurrentIconPack: String) : BaseAdapter() {
        internal var supportedPackages: ArrayList<IconPackInfo> = ArrayList(supportedPackages.values)
        internal var layoutInflater: LayoutInflater = LayoutInflater.from(context)

        init {
            Collections.sort(this.supportedPackages) { lhs, rhs -> lhs.label.toString().compareTo(rhs.label.toString(), ignoreCase = true) }

            val defaultLabel = "None"
            val icon = Utilities.getMyIcon(context)
            this.supportedPackages.add(0, IconPackInfo(defaultLabel, icon, ""))
        }

        override fun getCount(): Int {
            return supportedPackages.size
        }

        override fun getItem(position: Int): String {
            return supportedPackages[position].packageName
        }

        override fun getItemId(position: Int): Long {
            return 0
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view = convertView ?: layoutInflater.inflate(R.layout.iconpack_dialog, parent, false)
            val info = supportedPackages[position]
            val txtView = view.findViewById<TextView>(R.id.title)
            txtView.text = info.label
            val imgView = view.findViewById<ImageView>(R.id.icon)
            imgView.setImageDrawable(info.icon)
            val radioButton = view.findViewById<RadioButton>(R.id.radio)
            radioButton.isChecked = info.packageName == mCurrentIconPack
            return view
        }
    }

}

